### ISSUE TITLE
"Unresolved reference: retry" when plugin used inside custom plugin

### ISSUE THREAD

**Issue by thahnen** [gradle/test-retry-gradle-plugin#89]
I'm currently building a custom plugin for our gradle build system. This one uses the test-retry plugin on Test tasks.
The exact plugin was prior used inside the projects build.gradle.kts which was rather ugly so the idea came to migrate the working plugin to a custom plugin which can be used directly using a maven nexus.

My problem now ist the following: When using the retry function, IntelliJ can not find it while the plugin is developed in a specific kotlin file while doing so in the build.gradle.kts directly it works. Maybe I am missing sth. but i tried a lot.

The **build.gradle.kts** file for the custom plugin looks like this:
```kotlin
/** 1) Plugins used globally */
plugins {
    `java-gradle-plugin`
    `kotlin-dsl`
    `maven-publish`

    // external plugins
    //id("org.jetbrains.kotlin.jvm") version "1.3.72"
    id("org.jetbrains.kotlin.jvm") version "1.4.30-M1"
    id("org.gradle.test-retry") version "1.2.0" apply(false)
}


/** 2) General information regarding the plugin */
group   = property("plugin.group")!!
version = property("plugin.version")!!


/** 3) Apply following plugins to this custom plugin */
apply {
    plugin("kotlin")
    plugin("java-gradle-plugin")
    plugin("org.gradle.test-retry")
}


/** 4) Dependency source configuration */
repositories {
    maven(url = property("maven.nexus")!!)
    jcenter()
}


/** 5) Plugin dependencies */
dependencies {
    implementation("org.jetbrains.kotlin:kotlin-bom")
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
    implementation("org.gradle:test-retry-gradle-plugin:1.2.0")

    // used for testing
    testImplementation("org.jetbrains.kotlin:kotlin-test")
    testImplementation("org.jetbrains.kotlin:kotlin-test-junit")
    testImplementation(gradleTestKit())
}


/** 6) Configuration where to publish the plugin */
publishing {
    repositories {
        maven {
            url     = uri(property("maven.nexus")!!)
            name    = property("maven.nexus.name")!! as String

            // publishing credentials
            credentials {
                username = property("maven.nexus.login.username")!! as String
                password = property("maven.nexus.login.password")!! as String
            }
        }
    }
}


/** 7) Create new plugin */
gradlePlugin {
    plugins {
        create("ConfigureJava") {
            id                  = property("plugin.id")!! as String
            implementationClass = property("plugin.class")!! as String
        }
    }
}

```

There I could easily use "*import org.gradle.kotlin.dsl.retry*" but see the following **ConfigureJavaPlugin.kt** where I can not import it because it can not be found (weird):
```kotlin
import org.gradle.api.JavaVersion
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.JavaPluginConvention
import org.gradle.api.tasks.testing.Test

import org.gradle.kotlin.dsl.apply
import org.gradle.kotlin.dsl.create
import org.gradle.kotlin.dsl.configure
import org.gradle.kotlin.dsl.withType


/**
 *  ConfigureJavaPlugin:
 *  ===================
 *
 *  @author Tobias Hahnen
 *
 *  Plugin to configure Java and its gradle plugin dependencies before building using centralized settings.
 */
open class ConfigureJavaPlugin : Plugin<Project> {
    override fun apply(target: Project) {
        // apply Java / Test-Retry plugins
        target.apply(plugin = "java")
        target.apply(plugin = "org.gradle.test-retry")

        // load extension data
        val extension = target.extensions.create<ConfigureJavaPluginExtension>("configurejavaplugin")

        // INFO: To use extension afterEvaluate is necessary!
        target.afterEvaluate {
            // configures the basic Java plugin
            target.configure<JavaPluginConvention> {
                sourceCompatibility = extension.sourceCompatibility
                targetCompatibility = extension.targetCompatibility
            }

            // configure tasks of type "Test"
            target.tasks.withType<Test> {
                useJUnitPlatform()

                maxParallelForks    = extension.maxParallelForks
                maxHeapSize         = extension.maxHeapSize
                ignoreFailures      = extension.ignoreFailures
                failFast            = extension.failFast

                this.filter { excludeTestsMatching("**Suite") }

                this.retry {
                    maxRetries.set(extension.maxRetries)
                    failOnPassedAfterRetry.set(extension.failOnPassedAfterRetry)
                    maxFailures.set(extension.maxFailures)
                }
            }
        }
    }
}


/**
 *  ConfigureJavaPluginExtension:
 *  ============================
 *
 *  @author Tobias Hahnen
 *
 *  Extension that belongs to the same name plugin used for optional configuration.
 */
open class ConfigureJavaPluginExtension {
    var sourceCompatibility = JavaVersion.VERSION_1_8
    var targetCompatibility = JavaVersion.VERSION_11
    var maxParallelForks = (Runtime.getRuntime().availableProcessors() / 2) - 1
    var maxHeapSize = "2G"
    var ignoreFailures = true
    var failFast = true
    var maxRetries = 10
    var failOnPassedAfterRetry = true
    var maxFailures = 10
}
```

I do believe I am missing sth. but I've build custom plugins which where constructed the same way for other external gradle plugins before which worked (eg. SonarQube, ...).

**Comment by bamboo** [comment]
Hi, @thahnen.

> There I could easily use "import org.gradle.kotlin.dsl.retry" but see the following ConfigureJavaPlugin.kt where I can not import it because it can not be found 

That happens because while the Kotlin DSL can generate [type-safe accessors](https://docs.gradle.org/current/userguide/kotlin_dsl.html#type-safe-accessors) for extensions contributed by plugins applied in the `plugins` block of a Kotlin script (like the `retry` extension contributed by the `org.gradle.test-retry` plugin), it cannot do the same for plugins implemented as classes.

The Kotlin DSL provides the `configure` utility to facilitate the use of extensions in the absence of type-safe accessors: 

```kotlin
// configure tasks of type "Test"
target.tasks.withType<Test> {
    // configure the `retry` extension
    configure<org.gradle.testretry.TestRetryTaskExtension> {

    }
}
```

Alternatively, you could implement your custom plugin as a [precompiled script plugin](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins) and get the accessors generated for you.

I hope that helps.

**Comment by marcphilipp** [comment]
@bamboo AFAIK the `retry` method is _not_ an accessor generated by the Kotlin DSL since it's a task, not a project extension.

It's shipped as part of the plugin:
https://github.com/gradle/test-retry-gradle-plugin/blob/77597221329e261884e564ca77ba48a2ed69c224/plugin/src/main/kotlin/org/gradle/kotlin/dsl/testRetry.kt#L26-L27

**Comment by bamboo** [comment]
Thanks, @marcphilipp, I stand corrected!

I'm not sure what's going on there then. The following works for me with Gradle 6.8:

```kotlin
// build.gradle.kts
plugins {
    `kotlin-dsl`
}

repositories {
    gradlePluginPortal()
}

dependencies {
    implementation("org.gradle:test-retry-gradle-plugin:1.2.0")
}

// src/main/kotlin/plugins/CustomPlugin.kt
package plugins

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.tasks.testing.Test
import org.gradle.kotlin.dsl.retry
import org.gradle.kotlin.dsl.withType

class CustomPlugin : Plugin<Project> {
    override fun apply(target: Project) {
        target.tasks.withType<Test>().configureEach {
            retry {

            }
        }
    }
}
```

**Comment by thahnen** [comment]
Thanks to both of you! 👍 

I think one missing piece was the
```kotlin
repositories {
    gradlePluginPortal()
}
```
which I saw missing in my code and adjusted it. Therefor @bamboo is correct and it works until this point, because the options set in my retry-scope:
```kotlin
// ...

target.retry {
    maxRetries.set(extension.maxRetries)
    failOnPassedAfterRetry.set(extension.failOnPassedAfterRetry)
    maxFailures.set(extension.maxFailures)
}

// ...
```

are not found. As I investigated they are members of the TestRetryTaskExtension provided by this plugin. But this class, or interface i guess, can not be found (org.gradle.test-retry.TestRetryTaskExtension), I tried your other idea with the configuration of the plugin extension before where it wasn't found as well.

I did not take a look at the precompiled script plugin yet but it looks worth a try if this does not work.

**Comment by thahnen** [comment]
Time to close this issue.

I resolved it, it seemed to be an error inside IntelliJ IDEA that the classes inside the plugin could not be found and therefore refused to even build it. On the other hand I changed the gradle version multiple times while always invalidating IntelliJs internal cache.

With gradle version 6.7 and the newest IntelliJ IDEA build (version 2020.3) it now works with the small restriction that the classes won't be found but running gradle works now.
I suspect an error inside IntelliSense but I have no clue or knowledge regarding IDEA other than how to use it :D

Therefore it was my misconception assuming the error was this plugins fault.
